nilearn/regions/hierarchical_kmeans_clustering.py

```diff
@@ -0,0 +1,276 @@
+        If scaling is True, each cluster is scaled by the square root of its
+        size during transform(), preserving the l2-norm of the image.
+        inverse_transform() will apply inversed scaling to yield an image with
+        same l2-norm as input.
+
+    verbose: int, optional (default 0)
+        Verbosity level.
+
+    Attributes
+    ----------
+    `labels_ `: ndarray, shape = [n_features]
+        cluster labels for each feature.
+
+    `sizes_`: ndarray, shape = [n_features]
+        It contains the size of each cluster.
+
+    """
+
+    def __init__(self, n_clusters, init="k-means++", batch_size=1000,
+                 n_init=10, max_no_improvement=10, verbose=0, random_state=0,
+                 scaling=False):
+        self.n_clusters = n_clusters
+        self.init = init
+        self.batch_size = batch_size
+        self.n_init = n_init
+        self.max_no_improvement = max_no_improvement
+        self.verbose = verbose
+        self.random_state = random_state
+        self.scaling = scaling
+
+    def fit(self, X, y=None):
+        """Compute clustering of the data.
+
+        Parameters
+        ----------
+        X: ndarray, shape = [n_features, n_samples]
+            Training data.
+        y: Ignored
+
+        Returns
+        -------
+        self
+        """
+
+        X = check_array(X, ensure_min_features=2, ensure_min_samples=2,
+                        estimator=self)
+        n_features = X.shape[1]
+
+        if self.n_clusters <= 0:
+            raise ValueError("n_clusters should be an integer greater than 0."
+                             " %s was provided." % str(self.n_clusters))
+
+        if self.n_clusters > n_features:
+            self.n_clusters = n_features
+            warnings.warn("n_clusters should be at most the number of "
+                          "features. Taking n_clusters = %s instead."
+                          % str(n_features))
+        self.labels_ = hierarchical_k_means(X, self.n_clusters, self.init,
+                                            self.batch_size, self.n_init,
+                                            self.max_no_improvement,
+                                            self.verbose, self.random_state)
+        sizes = np.bincount(self.labels_)
+        sizes = sizes[sizes > 0]
+        self.sizes_ = sizes
+        self.n_clusters = len(np.unique(self.labels_))
+        return self
+
+    def transform(self, X, y=None):
+        """Apply clustering, reduce the dimensionality of the data.
+
+        Parameters
+        ----------
+        X: ndarray, shape = [n_features, n_samples]
+            Data to transform with the fitted clustering.
+
+        Returns
+        -------
+        X_red: ndarray, shape = [n_clusters, n_samples]
+            Data reduced with agglomerated signal for each cluster
+        """
+
+        check_is_fitted(self, "labels_")
+        unique_labels = np.unique(self.labels_)
+
+        mean_cluster = []
```

Comment:
you can use `np.empty` and fill the array if you don't want to build a list and then an array